pkg/virt-controller/network/ipamclaims.go

```diff
@@ -0,0 +1,207 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	virtv1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/network/namescheme"
+	"kubevirt.io/kubevirt/pkg/network/vmispec"
+
+	"kubevirt.io/client-go/kubecli"
+
+	ipamclaims "github.com/k8snetworkplumbingwg/ipamclaims/pkg/crd/ipamclaims/v1alpha1"
+	networkv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+)
+
+type IPAMClaimsManager struct {
+	client kubecli.KubevirtClient
+}
+
+func NewIPAMClaimsManager(client kubecli.KubevirtClient) *IPAMClaimsManager {
+	return &IPAMClaimsManager{
+		client: client,
+	}
+}
+
+type IPAMClaimParams struct {
+	ClaimName   string
+	NetworkName string
+}
+
+func (m *IPAMClaimsManager) CreateIPAMClaims(namespace string, vmiName string, interfaces []virtv1.Interface, networks []virtv1.Network, ownerRef *v1.OwnerReference) error {
+	nonAbsentNetworks := filterNonAbsentNetworks(interfaces, networks)
+	networkToIPAMClaimParams, err := m.GetNetworkToIPAMClaimParams(namespace, vmiName, nonAbsentNetworks)
+	if err != nil {
+		return fmt.Errorf("failed composing networkToIPAMClaimName: %w", err)
+	}
+
+	claims := composeIPAMClaims(namespace, ownerRef, networkToIPAMClaimParams)
+	err = m.createIPAMClaims(namespace, claims)
+	if err != nil {
+		return fmt.Errorf("failed IPAMClaims creation for VMI %s: %w", vmiName, err)
+	}
+
+	return nil
+}
+
+func composeIPAMClaims(namespace string, ownerRef *v1.OwnerReference, networkToIPAMClaimParams map[string]IPAMClaimParams) []*ipamclaims.IPAMClaim {
+	claims := []*ipamclaims.IPAMClaim{}
+	for netName, ipamClaimParams := range networkToIPAMClaimParams {
+		claims = append(claims, composeIPAMClaim(
+			namespace,
+			*ownerRef,
+			ipamClaimParams,
+			namescheme.GenerateHashedInterfaceName(netName),
+		))
+	}
+
+	return claims
+}
+
+func (m *IPAMClaimsManager) createIPAMClaims(namespace string, claims []*ipamclaims.IPAMClaim) error {
```

Comment:
imo since we call this just once, it doesnt merit a function
also about naming, lets please get other opinions,
since this function does create, this is its main verb, if exists it ensure it is the right one
(`ensureValidIPAMClaimForVMI` was renamed as such by Miguel request, in order to try implying what it does instead of a comment)

```
		if err != nil {
			if !k8serrors.IsAlreadyExists(err) {
				return fmt.Errorf("failed to create IPAMClaim: %w", err)
			}

			err = m.ensureValidIPAMClaimForVMI(namespace, claim.Name, claim.OwnerReferences[0].UID)
			if err != nil {
				return fmt.Errorf("failed validating IPAMClaim: %w", err)
			}
		}
```